File: node/stati_d.go
package main

import (
    "github.com/bradfitz/gomemcache/memcache"
    "fmt"
    "net/http"
    "io/ioutil"
    "encoding/json"
    "encoding/binary"
    "net"
    "bytes"
    "strconv"
    "time"
)

const (
    //URL_APPIDS = "http://222.186.49.42:3737/getallreleaseapp/?new"
    URL_APPIDS = "http://127.0.0.1:3737/getallreleaseapp/?new"
    HOST_MEMCACHED = "127.0.0.1:11211"
    HOST_STATI = "58.215.46.92:16410"
    LEN_HEAD = 4
    CACHE_INTERVAL = time.Second * 60
)

type Appid uint

type ReqStati struct {
    Op      string  `json:"op"`
    Appids  []Appid `json:"appids"`
    Topn    uint    `json:"topn"`
    Expand  string  `json:"expand"`
}

type RepStati struct {
    Op      string `json:"op"` 
    Expand  string `json:"expand"`
    Apps    []Rank `json:"apps"`
}

type Rank struct{
    Appid   uint            `json:"appid"`
    Sids    []RankDetail    `json:"sids"`
}

type RankDetail struct{
    Sid         uint    `json:"sid"`
    Usercount   uint    `json:"usercount"`
}

var mc *memcache.Client


func init() {
    mc = memcache.New(HOST_MEMCACHED)
}

func getAppids() []Appid {
    rep, _ := http.Get(URL_APPIDS)
    body, _ := ioutil.ReadAll(rep.Body)
    defer rep.Body.Close()

    var appids []Appid
    json.Unmarshal(body, &appids)    
    return appids
}


func getAppTopnSids() (rep RepStati) {
    conn, err := net.Dial("tcp", HOST_STATI)
    if err != nil {
        panic(err)
    }
    req := ReqStati{"batch_get_app_topn_sids", getAppids(), 12, ""} 
    jn, _ := json.Marshal(req)
    //head := make([]byte, LEN_HEAD)
    ////copy(head, string(len(jn)))
    //binary.LittleEndian.PutUint32(head, uint32(len(jn)))
    //msg := append(head, jn...)
    msg := make([]byte, len(jn) + LEN_HEAD)
    binary.LittleEndian.PutUint32(msg, uint32(len(jn)))
    copy(msg[LEN_HEAD:], jn)
    conn.Write(msg)

    var data bytes.Buffer
    //var rep RepStati
    buf := make([]byte, 2048)
    for {
        n, _ := conn.Read(buf)
        if n == 0 {
            break
        }
        data.Write(buf[:n])
        if data.Len() < LEN_HEAD {
            continue
        }
        len_msg := binary.LittleEndian.Uint32(data.Bytes()[:LEN_HEAD])
        if uint32(data.Len()) < len_msg + LEN_HEAD {
            continue
        }
        body := data.Bytes()[LEN_HEAD:len_msg+LEN_HEAD]
        json.Unmarshal(body, &rep)
        fmt.Println("result", rep.Apps)
        //return rep
        break
    }
    return
}

func cacheTopn() {
    defer func() {
        if r := recover(); r != nil {
            fmt.Println("recovered")
        }
    }()
    ranks := getAppTopnSids().Apps
    var top1s  []Rank
    for _, rank := range ranks {
        appid := rank.Appid
        top12, _ := json.Marshal([]Rank{rank})
        top1 := Rank{appid, []RankDetail{rank.Sids[0]}}
        top1s = append(top1s, top1)
        mc.Set(&memcache.Item{Key:strconv.Itoa(int(appid)), Value:top12})
    }
    jn, _ := json.Marshal(top1s)
    mc.Set(&memcache.Item{Key:"top1s", Value:jn})

}


func main() {
    for {
        cacheTopn()
        time.Sleep(CACHE_INTERVAL)
    }
}

File: node/stati_d.py
# -*- coding: utf-8 -*-

# django setting
import os, sys
path = os.path.abspath("..")
sys.path.insert(0, path)
os.environ['DJANGO_SETTINGS_MODULE'] = 'web.settings'

import time, memcache, json, urllib, random, syslog
from node2statistic import getAppTopnSids, mc

syslog.openlog("broker-stati-d")

APPIDS = (101213, 100767, 100877, 103350, 102850, 100907, 100905, 100824, 100815, 100492)
TOPN = 12   # 排名数
APPIDS_CACHE_TIME = 60 * 5
INTERVAL = 60

#BROKER_HOST = ("http://222.88.95.242:3737/", "http://121.14.39.131:3737/")
BROKER_HOST = ("http://127.0.0.1/", )
SUBFIX = "getallreleaseapp/?new"


def getAppids():
    appids = mc.get("APPIDS")
    if not appids:
        try:
            host = random.choice(BROKER_HOST)
            url = host + SUBFIX
            u = urllib.urlopen(url)
            appids = json.loads(u.read())
            mc.set("APPIDS", appids, APPIDS_CACHE_TIME)
            syslog.syslog("get from url")
        except:
            appids = APPIDS
            syslog.syslog("get from const")
    else:
        syslog.syslog("get from cache")
    return appids


# 兼容格式 [{"sids":[{"usercount":1, "sid":1}...], "appid":appid}...]
def cacheTopn():
    ret = getAppTopnSids(getAppids(), TOPN)
    top1s = []
    for r in ret:
        appid = r["appid"]
        sids = r["sids"]
        # 单个应用的前12名
        top12 = json.dumps([{"sids":sids, "appid":appid}])
        # 所有上线应用的第一名
        top1s.append({"sids":[sids[0]] if sids else [], "appid":appid})
        mc.set(str(appid), top12)
    mc.set("top1s", json.dumps(top1s))


import daemon
with daemon.DaemonContext(stdout=sys.stdout, stderr=sys.stderr):
    while True:
        cacheTopn()
        time.sleep(INTERVAL)

